File: src/ast.jl
"""
AST for phase-0 `ad`. Deliberately small — later phases add node types; phase 4 compiles
these to an interaction net rather than changing this shape. See `docs/architecture.md`.
"""
module AST

export Node, NumLit, Var, BackslashRef, BinOp, UnOp, Assign, Seq

abstract type Node end

struct NumLit <: Node
    value::String   # raw literal text; parsed to a Num.AdNum during eval
end

struct Var <: Node
    name::Char
end

"""
Reference to a `\\`-name (e.g. `\\pi`). Phase 0 seeds the lexer's name table but binds none
of them, so evaluating one of these always raises an unbound-name error -- see docs/language.md.
"""
struct BackslashRef <: Node
    name::String
end

struct BinOp <: Node
    op::Symbol       # :+, :-, :*, :/, :^
    lhs::Node
    rhs::Node
end

struct UnOp <: Node
    op::Symbol       # :-
    operand::Node
end

"`force` distinguishes `:=` (true) from `=` (false, assign-or-check)."
struct Assign <: Node
    name::Char
    force::Bool
    value::Node
end

struct Seq <: Node
    statements::Vector{Node}
end

end # module

File: src/ast.rs
//! The AST: a typed Rust enum, not an s-expression list. The Common Lisp attempt used
//! s-expressions on the reasoning that a later symbolic-algebra phase (`\expr`/`\eval`/
//! `\body`) could quote and rewrite the AST with ordinary list operations — but that's a
//! Lisp-ism, and doesn't transfer: Rust enums quote and rewrite fine via match-and-rebuild.
//!
//! Every node carries its own `Span`, tagged at construction time by the parser (see
//! `parser.rs` for why *at construction*, not on the way out of each `parse_*` call,
//! matters for parenthesized expressions).

use crate::span::Span;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Pow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnOp {
    Neg,
}

#[derive(Debug, Clone)]
pub struct Node {
    pub kind: NodeKind,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub enum NodeKind {
    NumLit(String),
    Var(char),
    BackslashRef(String),
    BinOp { op: BinOp, lhs: Box<Node>, rhs: Box<Node> },
    UnOp { op: UnOp, operand: Box<Node> },
    Assign { name: char, force: bool, value: Box<Node> },
    Seq(Vec<Node>),
}

impl Node {
    pub fn new(kind: NodeKind, span: Span) -> Self {
        Node { kind, span }
    }
}
